### PR TITLE
Support TypeScript 4.1 via typescript-estree

### LINKED ISSUE #9234: Support TypeScript 4.1
[TypeScript 4.1 beta has been released!](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1-beta/)

- typescript-eslint: https://github.com/typescript-eslint/typescript-eslint/issues/2583

### Template Literal Types
https://devblogs.microsoft.com/typescript/announcing-typescript-4-1-beta/#template-literal-types
```ts
type Greeting = `hello ${World}`;
```

### Key Remapping in Mapped Types
https://devblogs.microsoft.com/typescript/announcing-typescript-4-1-beta/#key-remapping-mapped-types
```ts
type MappedTypeWithNewKeys<T> = {
    [K in keyof T as NewKeyType]: T[K]
}
```

### PR DESCRIPTION
**Pull request by sosukesuzuki**
Partial of #9234

~~This PR tries [the approach](https://github.com/prettier/prettier/issues/9549#issuecomment-724768202) @thorn0 suggested (27e7b46)~~

- [x] I’ve added tests to confirm my change works.
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-graphql/printer-graphql.js [hunk 1]
```diff
@@ -1,15 +1,8 @@
...
-  join,
-  hardline,
-  line,
-  softline,
-  group,
-  indent,
-  ifBreak,
-} = require("../document").builders;
+  builders: { concat, join, hardline, line, softline, group, indent, ifBreak },
+} = require("../document");
```
Comment (sosukesuzuki) [inline src/language-graphql/printer-graphql.js:5]: I don't know why, but this change removes the TypeScript typecheck error. Why...?

**Comment by sosukesuzuki** [comment]
All prod tests are failed. I'm looking into that.

File: src/language-graphql/printer-graphql.js (hunk 1 above)
Comment (fisker) [reply src/language-graphql/printer-graphql.js:5]: Maybe https://github.com/prettier/prettier/pull/9207

**Comment by fisker** [comment]
Seems we can access decorators via `services`


```js
require("@typescript-eslint/typescript-estree").parseAndGenerateServices('@foo\nfunction bar(){}').services.program.getSourceFile().statements[0]
```

**Comment by fisker** [comment]
I'm going to take a shot

**Comment by fisker** [comment]
I tried to traverse ESTree Ast , get `TSNode` from `ESNode`, and check `decorators`, it didn't work, because `esTreeNodeToTSNodeMap.get(ClassBody)` (maybe the same as `esTreeNodeToTSNodeMap.get(ClassDeclaration)`) also has `decorators`.

File: src/language-js/postprocess.js [hunk 2]
```diff
@@ -129,6 +172,19 @@ function postprocess(ast, options) {
           node.optional = true;
         }
         break;
+      case "TSTemplateLiteralType": {
```
Comment (fisker) [inline src/language-js/postprocess.js:175]: I think we should print this, not transform it.

Comment (sosukesuzuki) [reply src/language-js/postprocess.js:175]: I have no strong opinion. Out of curiosity, why?

Comment (fisker) [reply src/language-js/postprocess.js:175]: Maybe we'll support `formatAST()`

File: src/language-js/parser-typescript.js [hunk 3]
```diff
@@ -30,12 +30,16 @@ function parse(text, parsers, opts) {
...
+    ...opts,
+    originalText: text,
+    tsParseResult: result,
+  });
 }
 
 function tryParseTypeScript(text, jsx) {
   const parser = require("@typescript-eslint/typescript-estree");
-  return parser.parse(text, {
+  return parser.parseAndGenerateServices(text, {
```
Comment (thorn0) [inline src/language-js/parser-typescript.js:42]: I wonder about the performance impact of this change.

Comment (fisker) [reply src/language-js/parser-typescript.js:42]: I haven't check, but I think they generate services anyway.

@bradzacher Can you tell us the difference?

Comment (fisker) [reply src/language-js/parser-typescript.js:42]: Maybe we can use parse if no `@` ?

File: src/language-js/utils.js [hunk 4]
```diff
@@ -464,6 +464,7 @@ const simpleTypeAnnotations = new Set([
   "BigIntLiteralTypeAnnotation",
   "NumberLiteralTypeAnnotation",
   "TSLiteralType",
+  "TSTemplateLiteralType",
```
Comment (thorn0) [inline src/language-js/utils.js:467]: Not sure about this. Unlike other node types in this list, template literal types aren't that simple and can contain complex expressions. We can keep them here though and see if users will complain about it. Probably it's okay.

Comment (fisker) [reply src/language-js/utils.js:467]: That's also what I thought 😄

File: src/language-js/parser-typescript.js (hunk 3 above)
Comment (bradzacher) [reply src/language-js/parser-typescript.js:42]: without providing `parserOptions.project` it shouldn't be too different I'd have thought.

If you call `parse`, we'll create a `sourceFile`, convert it and that's it.
If you call `parseAndGenerateServices`, we'll create a TS program with one file in isolated mode, parse it and convert it.

There'll be a small overhead as TS will create some data-structures behind the scenes to enable you to access type information - but as all type information is computed lazily it shouldn't be an issue.

We could always expose a new API like `parseWithNodeMaps` which is like `parse` except it returns the AST + node maps.

Comment (bradzacher) [reply src/language-js/parser-typescript.js:42]: https://github.com/typescript-eslint/typescript-eslint/pull/2760

I will merge this and it'll be available with our release next monday

Comment (fisker) [reply src/language-js/parser-typescript.js:42]: Great! Thank  you!  @bradzacher

**Comment by fisker** [comment]
@typescript-eslint/typescript-estree@4.8.0 released, working on it.

**Comment by fisker** [comment]
@bradzacher Did I do something wrong? There is no info in `esTreeNodeToTSNodeMap`


```js
> const {ast, esTreeNodeToTSNodeMap} = require("@typescript-eslint/typescript-estree").parseWithNodeMaps('@foo()\nfunction bar(){}')
undefined
> esTreeNodeToTSNodeMap.get(ast.body[0])
undefined
```


```js
> const {ast, services:{esTreeNodeToTSNodeMap}} = require("@typescript-eslint/typescript-estree").parseAndGenerateServices('@foo()\nfunction bar(){}')
undefined
> esTreeNodeToTSNodeMap.get(ast.body[0])
<ref *1> NodeObject {
  pos: 0,
  end: 23,
  flags: 0,
  modifierFlagsCache: 0,
  transformFlags: 1050625,
  parent: <ref *2> SourceFileObject
//...
```

**Comment by bradzacher** [comment]
FFS I'm a stooge.
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/parser.ts#L383

The `false` here means "don't calculate maps".
I can't believe I didn't think about that!
My test just asserted the maps existed, not that they contained anything.
🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦 

I'll have a fix released ASAP.

**Comment by fisker** [comment]
Found problem, https://github.com/typescript-eslint/typescript-eslint/blob/3333040f02d14531efc1ad70af7dbd1e6a24c1fa/packages/typescript-estree/src/parser.ts#L383 third argument should be `true`

**Comment by bradzacher** [comment]
https://github.com/typescript-eslint/typescript-eslint/pull/2773 is merged, but I can't do an NPM release right now as azure pipelines have an outage....
Yeah we need to move it to GH actions. Will ping back when it's released.

For now you can test against
https://www.npmjs.com/package/@typescript-eslint/typescript-estree/v/4.8.1-alpha.1
Which releases via GH actions.

**Review by fisker — APPROVED** [review]

**Review by thorn0 — APPROVED** [review]

**Comment by sosukesuzuki** [comment]
Can we merge this as is? or should wait `4.8.1`?

**Comment by fisker** [comment]
I don't think we need wait for @typescript-eslint/typescript-estree , but do we need wait for typescript?

**Comment by sosukesuzuki** [comment]
TypeScript 4.1 will seem be released soon according to https://github.com/microsoft/TypeScript/issues/40124. Let's wait.

**Comment by bradzacher** [comment]
4.8.1 is live with the fix :)

**Comment by fisker** [comment]
> Hey all, we're going to retarget to Thursday (11/19) just to ensure that everything is working as optimally as possible and documented.

https://github.com/microsoft/TypeScript/issues/40124#issuecomment-729289818

**Comment by sosukesuzuki** [comment]
I see. let's merge this PR for now. I'll do the other work required for the release.

When TypeScript 4.1 is released, I'll create a PR just to update it.

**Pull request merged**